File: implementations/0.2/1-hybrid/src/pinpoint/paths.py
from pathlib import PurePosixPath

from pinpoint.models import ROOT_DIRS


def _pad2(val: str) -> str:
    return val.zfill(2) if val.isdigit() else val


def _get(tags: dict, key: str) -> str:
    vals = tags.get(key, [])
    return vals[0] if vals else ""


def derive_path(root: str, tags: dict, original_filename: str) -> str:
    ext = PurePosixPath(original_filename).suffix
    root_dir = ROOT_DIRS.get(root, root)
    name = _get(tags, "name")

    if root == "memory":
        month = _get(tags, "month") or "unknown"
        event = _get(tags, "event")
        parts = [root_dir, month]
        if event:
            parts.extend(event.split(":"))
        if name:
            parts.append(f"{name}{ext}")
        else:
            parts.append(original_filename)
        return "/".join(parts)

    if root == "music":
        artist = _get(tags, "artist") or "_unknown"
        album = _get(tags, "album")
        year = _get(tags, "year")
        track = _get(tags, "track")

        parts = [root_dir, artist]

        if album:
            album_dir = f"[{year}] {album}" if year else album
            parts.append(album_dir)

        if name:
            if track:
                parts.append(f"{_pad2(track)} - {name}{ext}")
            else:
                parts.append(f"{name}{ext}")
        else:
            if track:
                parts.append(f"{_pad2(track)} - {original_filename}")
            else:
                parts.append(original_filename)
        return "/".join(parts)

    if root == "movie":
        series = _get(tags, "series")
        year = _get(tags, "year")
        parts = [root_dir]
        if series:
            parts.extend(series.split(":"))
        if name:
            fname = f"{name} [{year}]{ext}" if year else f"{name}{ext}"
            parts.append(fname)
        else:
            parts.append(original_filename)
        return "/".join(parts)

    if root == "tv":
        show = _get(tags, "show") or "_unknown"
        season = _get(tags, "season")
        episode = _get(tags, "episode")
        parts = [root_dir, show]
        if season:
            parts.append(f"Season {_pad2(season)}")
        if name:
            if episode:
                parts.append(f"{_pad2(episode)} - {name}{ext}")
            else:
                parts.append(f"{name}{ext}")
        else:
            if episode:
                parts.append(f"{_pad2(episode)}{ext}")
            else:
                parts.append(original_filename)
        return "/".join(parts)

    if root == "podcast":
        show = _get(tags, "show") or "_unknown"
        episode = _get(tags, "episode")
        parts = [root_dir, show]
        if name:
            if episode:
                parts.append(f"{_pad2(episode)} - {name}{ext}")
            else:
                parts.append(f"{name}{ext}")
        else:
            if episode:
                parts.append(f"{_pad2(episode)} - {original_filename}")
            else:
                parts.append(original_filename)
        return "/".join(parts)

    if root == "book":
        author = _get(tags, "author") or "_unknown"
        series = _get(tags, "series")
        parts = [root_dir, author]
        if series:
            parts.extend(series.split(":"))
        if name:
            parts.append(f"{name}{ext}")
        else:
            parts.append(original_filename)
        return "/".join(parts)

    if root == "comedy":
        artist = _get(tags, "artist") or "_unknown"
        year = _get(tags, "year")
        parts = [root_dir, artist]
        if name:
            fname = f"[{year}] {name}{ext}" if year else f"{name}{ext}"
            parts.append(fname)
        else:
            parts.append(original_filename)
        return "/".join(parts)

    return f"{root_dir}/{original_filename}"

File: implementations/2-js/src/paths.js
import { extname } from "node:path";
import { ROOT_DIRS } from "./models.js";

function firstOr(tags, key, fallback = "_unknown") {
  const vals = tags[key];
  if (!vals || vals.length === 0) return fallback;
  return vals[0];
}

function has(tags, key) {
  const vals = tags[key];
  return vals && vals.length > 0 && vals[0] !== "";
}

function pad2(n) {
  const s = String(n).replace(/^0+/, "");
  return s.padStart(2, "0");
}

function yearFromDate(tags) {
  if (has(tags, "year")) return firstOr(tags, "year");
  const d = firstOr(tags, "date", "");
  if (d && d.length >= 4) return d.slice(0, 4);
  return "";
}

function monthFromDate(tags) {
  const d = firstOr(tags, "date", "");
  if (d && d.length >= 7) return d.slice(0, 7);
  return d ? d.slice(0, 4) + "-01" : "";
}

export function derivePath(root, tags, originalFilename) {
  const ext = extname(originalFilename);
  const stem = originalFilename.slice(0, -ext.length || undefined);
  const rootDir = ROOT_DIRS[root] || root;

  const name = has(tags, "name") ? firstOr(tags, "name") : stem;

  switch (root) {
    case "memory": {
      const month = monthFromDate(tags) || "_unknown";
      const event = firstOr(tags, "event", "");
      const eventSegments = event ? event.split(":").map((s) => s.trim()) : [];
      const parts = [rootDir, month, ...eventSegments];
      const filename = name + ext;
      return [...parts, filename].join("/");
    }

    case "music": {
      const artist = firstOr(tags, "artist");
      const album = firstOr(tags, "album", "");
      const track = firstOr(tags, "track", "");
      const year = yearFromDate(tags);

      const parts = [rootDir, artist];

      if (album) {
        const albumDir = year ? `[${year}] ${album}` : album;
        parts.push(albumDir);
      }

      let filename;
      if (track) {
        filename = `${pad2(track)} - ${name}${ext}`;
      } else {
        filename = `${name}${ext}`;
      }
      parts.push(filename);
      return parts.join("/");
    }

    case "movie": {
      const series = firstOr(tags, "series", "");
      const year = yearFromDate(tags);
      const parts = [rootDir];
      if (series) {
        series.split(":").forEach((s) => parts.push(s.trim()));
      }
      const filename = year ? `${name} [${year}]${ext}` : `${name}${ext}`;
      parts.push(filename);
      return parts.join("/");
    }

    case "tv": {
      const show = firstOr(tags, "show");
      const season = firstOr(tags, "season", "");
      const episode = firstOr(tags, "episode", "");
      const parts = [rootDir, show];
      if (season) parts.push(`Season ${pad2(season)}`);
      let filename;
      if (episode) {
        filename = `${pad2(episode)} - ${name}${ext}`;
      } else {
        filename = `${name}${ext}`;
      }
      parts.push(filename);
      return parts.join("/");
    }

    case "podcast": {
      const show = firstOr(tags, "show");
      const episode = firstOr(tags, "episode", "");
      const parts = [rootDir, show];
      let filename;
      if (episode) {
        filename = `${pad2(episode)} - ${name}${ext}`;
      } else {
        filename = `${name}${ext}`;
      }
      parts.push(filename);
      return parts.join("/");
    }

    case "book": {
      const author = firstOr(tags, "author");
      const series = firstOr(tags, "series", "");
      const parts = [rootDir, author];
      if (series) {
        series.split(":").forEach((s) => parts.push(s.trim()));
      }
      parts.push(`${name}${ext}`);
      return parts.join("/");
    }

    case "comedy": {
      const artist = firstOr(tags, "artist");
      const year = yearFromDate(tags);
      const parts = [rootDir, artist];
      const filename = year ? `[${year}] ${name}${ext}` : `${name}${ext}`;
      parts.push(filename);
      return parts.join("/");
    }

    default:
      return `${rootDir}/${originalFilename}`;
  }
}

export function resolveCollision(fullPath, db) {
  const ext = extname(fullPath);
  const base = fullPath.slice(0, -ext.length);
  let candidate = fullPath;
  let suffix = 1;
  while (
    db
      .prepare("SELECT 1 FROM files WHERE managed_path = ? AND status = 'managed'")
      .get(candidate)
  ) {
    candidate = `${base}-${suffix}${ext}`;
    suffix++;
  }
  return candidate;
}
